lxd/instance/drivers/driver_lxc.go

```diff
@@ -7913,8 +7938,46 @@ func (d *lxc) removeDiskDevices() error {
 	return nil
 }
 
-// Network I/O limits.
-func (d *lxc) setNetworkPriority() error {
+func setNetworkPriorityWithNetfilter(d *lxc, networkPriority int) error {
+	// in setNetworkPriorityWithCgroup we set skb->priority on all
+	// skb's that comes from the container cgroup on the enter to
+	// the host interfaces
+	// here we go another way, we set skb->priority on all skb's
+	// that exits the container network namespace once skb exits
+	// from the host peer of the container network device.
+
+	// Check that we at least succeeded to set an entry
+	success := false
+	var lastError error
+	for devName, devConfig := range d.ExpandedDevices() {
+		if devConfig["type"] != "nic" {
+			continue
+		}
+
+		hostDevName := d.ExpandedConfig()[fmt.Sprintf("volatile.%s.host_name", devName)]
+
+		var err error
+
+		err = d.state.Firewall.InstanceClearNetPrio(d.Project().Name, d.Name(), hostDevName)
+		if err == nil && networkPriority != 0 {
+			err = d.state.Firewall.InstanceSetupNetPrio(d.Project().Name, d.Name(), hostDevName, networkPriority)
+		}
+
+		if err == nil {
```

Comment:
If there are multiple devices with type `nic` and setting the net prio fails for all of them, will it only display the error message of the last device? 
Would it be possible to return here right away if it fails for one of the devices?

Reply:
Yes, of course. But I've just followed existing logic from `setNetworkPriorityWithCgroup` (previous name `setNetworkPriority`). As far as I understand the idea here is that failing netprio setting is not very critical and we are comfortable with the case when network priority was successfully set on at least one device.

Reply:
Just a design opinion, but I usually don't like nesting logic in `if err == nil` conditions, instead I go with the more conventional `if err != nil` pattern, but log the error and `continue` instead of returning the error. I find this reads more consistently, and has the added benefit of logging the additional errors in case they're of interest.

```go
NetworkPriorityWithNetfilter(d *lxc, networkPriority int) error {
    errs := make([]error, 0, len(d.ExpandedDevices())
    for devName, devConfig := range d.ExpandedDevices() {
        if devConfig["type"] != "nic" {
          continue
        }

        hostDevName := d.ExpandedConfig()[fmt.Sprintf("volatile.%s.host_name", devName)]

	err := d.state.Firewall.InstanceClearNetPrio(d.Project().Name, d.Name(), hostDevName)
        if err != nil {
            errs = append(errs, err)
            logger.Warn("Failed to clear instance network priority", logger.Ctx{"name": d.Name(), "project": d.Project().Name(), "error": err}
            
            continue
        }
       
        if networkPriority == 0 {
            continue
        }
 
        err = d.state.Firewall.InstanceSetupNetPrio(d.Project().Name, d.Name(), hostDevName, networkPriority)
        if err != nil {
            errs = append(errs, err)
            logger.Warn("Failed to setup instance network priority", logger.Ctx{"name": d.Name(), "project": d.Project().Name(), "error": err}

            continue
        }
   }

   if len(errs) > 0 {
      return fmt.Errorf("Failed to set network priority for %d devices. Last error: %w", len(eers), errs[len(errs) - 1])
   }

   return nil
}

```

Reply:
Ah shoot my condition at the end is wrong, It will fail if there are any errors. I'll add back the `succes` string :)

Reply:
One difference in the my implementation is that the `if !succes` case will potentially return a `nil` if `d.ExpandedDevices()` is empty. In your case: 

```go
if !success {
		return fmt.Errorf("Failed to set network device priority: %s", lastError)
	}
```

You create a new error, but `lastError` can still be `nil` if `d.ExpandedDevices()` is empty. Do we want this to error out if there are no devices, or just silently fail to set anything and move on?